STM32 mcu support the Vbat monitoring channel

FRASTM opened the pull request:
Adding the support of the Vbat monitoring internal channel on ADC
It includes a sample application (similar to the samples/sensor/stm32_temp_sensor) to demonstrate the Vbat monitoring

Valid on nucleo_g071rb as a first target, to be extended to any stm32 mcu with Vbat internal channel on ADC
For example, on the stm32G071, ADC 1 channel 14 is for internal VBat measurement with an resistor divider of 1/3
On stm32F401, the Vbat monitoring is available on ADC1 channel 18 with a bridge divider of 4.

requires https://github.com/zephyrproject-rtos/zephyr/pull/43861

**NOTE**
 this PR only configure the Vbat measurement for a nucleo_g071rb through a overlay in the sample
 A new PR must come to include the `has-vbat-channel; ` to the ADC node of any stm32  DTS


Fixes https://github.com/zephyrproject-rtos/zephyr/issues/41772

Signed-off-by: Francois Ramu [francois.ramu@st.com](mailto:francois.ramu@st.com)

FRASTM committed:
drivers: adc: stm32 internal channel for VBat measurement

It allows the path for VBATEN on stm32 series
to monitor the Vbat voltage, in case of CONFIG_STM32_VBAT
Add the common Vref value as a property of the ADC.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
dts: bindings: introduce a new stm32 vbat  monitor as a sensor

This new DTS Node is defining a ratio for Vbat internal
bridge of monitoring sensor connected
to a ADC internal channel. The voltage reference value
is given by the ADC of the stm32.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: sensor: add a stm32 sensor driver for the internal VBat

Similar to the internal temperature sensor of the stm32
this driver controls the Vbat monitoring in Volts,
using an ADC internal input and the stm32-vbat node of the DTS.
The ref voltage is given by the ADC.
Note that stm32F1x does not propose the feature.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
samples: sensor: new sample to demonstrate the Vbat monitoring

This new application is monitoring the Vbat in Volts
from the internal ADC.
Running on the nucleo_g071rb and nucle_wb55rg boards.
Vref is a property of the ADC.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
dts: bindings: stm32 adc has a Vbat channel monitoring

Like other temperature or vrefint sensor, the stm32 mcu
also have a Vbat monitoring internal channels on ADC.
Add this entry to the device tree.
The vref is usually 3300mV present on the target board.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
$ west build -p auto -b nucleo_g071rb samples/sensor/stm32_vbat_sensor/


```
*** Booting Zephyr OS build zephyr-v3.0.0-1258-gab2f5*
STM32 Vbat sensor test
Current Vbat voltage: 3345.0 V
Current Vbat voltage: 3342.0 V
...

```

ycsin commented on drivers/adc/adc_stm32.c:
```diff
@@ -993,7 +999,8 @@ static int adc_stm32_init(const struct device *dev)
 	LL_ADC_SetCommonPathInternalCh(__LL_ADC_COMMON_INSTANCE(adc),
-				       LL_ADC_PATH_INTERNAL_VREFINT);
+				       LL_ADC_PATH_INTERNAL_VREFINT |
+				       LL_ADC_PATH_INTERNAL_VBAT);
```
Why is this required? Isn't it handled by the `adc_stm32_setup_channels`?

FRASTM replied:
The one in the Init is required.
adc_stm32_setup_channels is not always called

ycsin replied:
This looks unnecessary in `adc_stm32_init`, user should call `adc_stm32_channel_setup` before `adc_read` which will setup the `LL_ADC_PATH_INTERNAL_VREFINT` or `LL_ADC_PATH_INTERNAL_VBAT` accordingly.

Setting `LL_ADC_PATH_INTERNAL_VBAT` here will connect the VBAT pin to the internal voltage divider on init, this will take away the special feature of the VBAT pin (to save power), as now the voltage divider will drain the battery.

I recommend that `LL_ADC_SetCommonPathInternalCh` things to be removed from the init function. Based on my testing, removing them doesn't seem to cause any issue.

ycsin commented on drivers/adc/adc_stm32.c:
```diff
@@ -811,6 +812,11 @@ static void adc_stm32_setup_channels(const struct device *dev, uint8_t channel_i
 		adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_VREFINT);
 	}
+
+	if (config->has_vbat_channel &&
```
The reference manual suggests that the link to be enabled only when vbat is being sampled. Do you mind implementing that in this PR as well?

FRASTM replied:
correct, this part could be under
`#if defined(CONFIG_STM32_VBAT)`

FRASTM replied:
plus  LL_ADC_SetCommonPathInternalChAdd  to set LL_ADC_PATH_INTERNAL_VBAT

FRASTM commented:
includes the https://github.com/zephyrproject-rtos/zephyr/pull/43861
Add flag `#if defined(CONFIG_STM32_VBAT)  `to set the VBAT internal divider

erwango commented on drivers/sensor/stm32_vbat/stm32_vbat.c:
```diff
@@ -0,0 +1,126 @@
+
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_vbat)
+#define DT_DRV_COMPAT st_stm32_vbat
+#define HAS_CALIBRATION 0
```
It doesn't seem to be used anywhere.

erwango commented on samples/sensor/stm32_vbat_sensor/README.rst:
```diff
@@ -0,0 +1,46 @@
+.. _stm32_vbat_sensor:
+
+STM32 VBat Sensor
+########################
```
Adjust number of #####

erwango commented on samples/sensor/stm32_vbat_sensor/README.rst:
```diff
@@ -0,0 +1,46 @@
+
+In order to run this sample, make sure to enable ``stm32_vbat`` node in your
+board DT file.
+Through a board overlay or so :
```
This wording is not clear

erwango commented on samples/sensor/stm32_vbat_sensor/README.rst:
```diff
@@ -0,0 +1,46 @@
+In order to run this sample, make sure to enable ``stm32_vbat`` node in your
+board DT file.
+Through a board overlay or so :
+/ {
+ stm32_vbat: stm32vbat {
+  compatible = "st,stm32-vbat";
+  label = "VBAT";
+  io-channels = <&adc1 14>;
+  ratio = 3;
+  status = "okay";
+ };
+};
```
Please use appropriate code snippets

erwango commented on samples/sensor/stm32_vbat_sensor/sample.yaml:
```diff
@@ -0,0 +1,9 @@
+    depends_on: adc
+    tags: sensors
+    build_only: true
+    platform_allow: nucleo_g071rb
```
Should it be build only ?
Can't we add a harness that checks voltage is valid ?
```
    harness: console
    timeout: 10
    harness_config:
      type: one_line
      regex:
        - "Current Vbat voltage: 33(.0) V"
```

erwango commented on samples/sensor/stm32_vbat_sensor/README.rst:
```diff
@@ -0,0 +1,46 @@
+
+.. code-block:: console
+
+Current Vbat voltage: 3348.0 V
```
mV

erwango commented on drivers/sensor/stm32_vbat/Kconfig:
```diff
@@ -0,0 +1,10 @@
+# STM32 temperature sensor configuration options
```
Battery voltage

erwango reviewed (changes requested):
One last comment.

Otherwise, for bisectibility, it would be nice to reorder the commits in the logical order:
-bindings
-sensor driver
-adc impact
-sample

ycsin commented on drivers/adc/adc_stm32.c:
```diff
@@ -811,6 +815,13 @@ static void adc_stm32_setup_channels(const struct device *dev, uint8_t channel_i
+	/* Enable the bridge divider only when needed for ADC conversion. */
+	if (config->has_vbat_channel &&
+		__LL_ADC_CHANNEL_TO_DECIMAL_NB(LL_ADC_CHANNEL_VBAT) == channel_id) {
+		adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_VBAT);
```
This line will set the common path for `LL_ADC_PATH_INTERNAL_VBAT`, which connects the VBAT pin to the internal voltage divider. According to the reference manual, the connection should only be made during the sampling, and VBAT should be disconnected from the voltage divider after sampling to avoid power drain.

I think something should call `LL_ADC_SetCommonPathInternalChRem()` at the end of `start_read` or in `adc_stm32_read` after `start_read` to remove `LL_ADC_PATH_INTERNAL_VBAT`

ycsin commented on drivers/sensor/stm32_vbat/stm32_vbat.c:
```diff
@@ -0,0 +1,126 @@
+
+	k_mutex_lock(&data->mutex, K_FOREVER);
+
+	rc = adc_read(data->adc, sp);
```
The driver should setup the channel (`adc_channel_setup`) to sample the VBAT before `adc_read`, Otherwise I guess this driver can't read its desired channel, if something else configured the channel after `stm32_vbat_init`

FRASTM replied:
that's the case :
```
	k_mutex_lock(&data->mutex, K_FOREVER);

	rc = adc_channel_setup(data->adc, &data->adc_cfg);

	rc = adc_read(data->adc, sp);
```

ycsin commented on drivers/sensor/stm32_vbat/stm32_vbat.c:
```diff
@@ -0,0 +1,126 @@
+		return -ENODEV;
+	}
+
+	*accp = (struct adc_channel_cfg){ .gain = ADC_GAIN_1,
+					  .reference = ADC_REF_INTERNAL,
+					  .acquisition_time = ADC_ACQ_TIME_MAX,
+					  .channel_id = data->channel,
+					  .differential = 0 };
```
Looks like this can be made into a constant

FRASTM replied:
which purpose, we have a code that could remain very similar to stm32_temp.c

FRASTM replied:
well from https://github.com/zephyrproject-rtos/zephyr/pull/44052 I see what you mean

ycsin commented on drivers/sensor/stm32_vbat/stm32_vbat.c:
```diff
@@ -0,0 +1,126 @@
+		.buffer = &data->sample_buffer,
+		.buffer_size = sizeof(data->sample_buffer),
+		.resolution = 12,
+		.calibrate = true,
```
Is the `calibrate` required? My testing shows that this makes no difference to the final result.

FRASTM replied:
no, removing

ycsin commented on drivers/adc/adc_stm32.c:
```diff
@@ -1079,6 +1096,7 @@ static const struct adc_driver_api api_stm32_driver_api = {
 #ifdef CONFIG_ADC_ASYNC
 	.read_async = adc_stm32_read_async,
 #endif
+	.ref_internal = STM32_ADC_VREF_MV, /* VREF is usually connected to VDD */
```
This line doesn't seem to be required to sample vbat voltage

ycsin replied:
Ok, its from another PR

FRASTM commented:
The sample here is defining an overlay with 
```
&adc1 {
    status = "okay";
    has-vbat-channel;
};
```
 but this would come in a next PR where ADC are updated with +    has-vbat-channel; in the DTS node.

FRASTM commented:
update the vbat driver in the same way as https://github.com/zephyrproject-rtos/zephyr/pull/44052, to allow other than vbat channel setup (out of the init)
reduce the sample application to display only one vbat measurement : no need to repeat endlessly

ycsin commented on samples/sensor/stm32_vbat_sensor/src/main.c:
```diff
@@ -0,0 +1,46 @@
+	}
+
+	printk("Current Vbat voltage: %.1f mV\n", sensor_value_to_double(&val));
+	k_sleep(K_MSEC(2000));
```
I think this line isn't required?

ycsin commented on samples/sensor/stm32_vbat_sensor/README.rst:
```diff
@@ -0,0 +1,52 @@
+
+.. code-block:: console
+
+Current Vbat voltage: 3348.0 mV
+Current Vbat voltage: 3345.0 mV
+Current Vbat voltage: 3342.0 mV
```
Update

ycsin commented:
25fd72f22c29d491a7766edde971edd2d2ec52ed & 7c9df1e939f3327d6d5f094c59bb678ef2236333 should be squashed?

ycsin commented:
@FRASTM I think it would be more interesting if the sample reads the VBAT in a loop, that way a user can varies the voltage and the terminal can display that in 'real-time'. Just my two cents

ycsin commented:
> update the vbat driver in the same way as #44052, to allow other than vbat channel setup (out of the init) reduce the sample application to display only one vbat measurement : no need to repeat endlessly

ycsin commented:
Clicked the wrong button, sorry!

ycsin commented on drivers/sensor/stm32_vbat/stm32_vbat.c:
```diff
@@ -0,0 +1,122 @@
+
+	k_mutex_lock(&data->mutex, K_FOREVER);
+
+	rc = adc_channel_setup(data->adc, &data->adc_cfg);
+	LOG_DBG("Setup AIN%u got %d", data->adc_cfg.channel_id, rc);
```
should return on error

```suggestion
	rc = adc_channel_setup(data->adc, &data->adc_cfg);
	if (rc) {
		LOG_DBG("Setup AIN%u got %d", data->adc_cfg.channel_id, rc);
		return rc;
	}
```

FRASTM replied:
ok

ycsin commented on drivers/sensor/stm32_vbat/stm32_vbat.c:
```diff
@@ -0,0 +1,122 @@
+
+	k_mutex_unlock(&data->mutex);
+
+	return 0;
```
should return  error code of `adc_read`

```suggestion
	return rc;
```

FRASTM replied:
ok

FRASTM commented:
rebase on 806cc70f7c6a5530b03c16cf0a0af3e34e59ad58
test the return value of `adc_channel_setup`

Not changed the sample application: not looping on vbat measurement on that sample ( _it is just a sample_) application : it is not likely to change on this nucleo target board (Vbat is fixed to VDD).

ycsin commented on drivers/sensor/stm32_vbat/stm32_vbat.c:
```diff
@@ -0,0 +1,126 @@
+
+	if (rc) {
+		LOG_DBG("Setup AIN%u got %d", data->adc_cfg.channel_id, rc);
+		return rc;
```
need to unlock mutex before returning

FRASTM replied:
changed

GeorgeCGV replied:
@FRASTM, but the mutex is not unlocked

ycsin reviewed (dismissed):
LGTM

erwango commented on drivers/sensor/stm32_vbat/stm32_vbat.c:
```diff
@@ -0,0 +1,127 @@
+
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_vbat)
+#define DT_DRV_COMPAT st_stm32_vbat
+#define HAS_CALIBRATION 0
```
Is that required ?

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -832,6 +833,14 @@ static void adc_stm32_setup_channels(const struct device *dev, uint8_t channel_i
 		__LL_ADC_CHANNEL_TO_DECIMAL_NB(LL_ADC_CHANNEL_VREFINT) == channel_id) {
 		adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_VREFINT);
 	}
+#if defined(CONFIG_STM32_VBAT)
```
Is that required ?
Obviously it will save few bytes, but can we imagine this being used w/o stm32_vbat sensor ?
Also if this to be kept, should we also use `STM32_TEMP` for the above temp related lines ?

FRASTM replied:
the LL_ADC_CHANNEL_VBAT is not available in all the stm32 mcus (unlike Temp and vRef )

erwango replied:
ok, then does it mean CONFIG_STM32_VBAT is not compatible with all series ?

FRASTM replied:
**stm32F1x** has no Vbat monitoring channel (but has a temperature sensor channel)

erwango commented on samples/sensor/stm32_vbat_sensor/README.rst:
```diff
@@ -0,0 +1,50 @@
+********
+
+This sample periodically reads the Vbat from the STM32 Internal
+Sensor and display the results.
```
displays

erwango commented on samples/sensor/stm32_vbat_sensor/boards/nucleo_g071rb.overlay:
```diff
@@ -0,0 +1,22 @@
+
+&adc1 {
+    status = "okay";
+    has-vbat-channel;
```
Please populate `has-vbat-channel` directly in dtsi files similarly to what was done for `has-temp-channel` and `has-vref-channel`

FRASTM replied:
ok

FRASTM committed:
dts: arm: stm32g0 has vbat capability on adc1

Like has-temp-channel or has-vref-channel the vbat property
on the channel 14 of the ADC1.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
rebase on 2a9b2dd7b950d3a324cdc58081539b8436c0bd41
change requested

ycsin commented on drivers/sensor/stm32_vbat/stm32_vbat.c:
```diff
@@ -0,0 +1,125 @@
+		.reference = ADC_REF_INTERNAL,
+		.acquisition_time = ADC_ACQ_TIME_MAX,
+		.channel_id = DT_INST_IO_CHANNELS_INPUT(0),
+		.differential = 0
```
nitpick, probably this line can end with a comma

FRASTM replied:
ok

ycsin reviewed (dismissed):
LGTM

erwango commented on samples/sensor/stm32_vbat_sensor/src/main.c:
```diff
@@ -0,0 +1,45 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <device.h>
+#include <drivers/sensor.h>
+#include <sys/printk.h>
```
```suggestion
#include <zephyr/zephyr.h>
#include <zephyr/device.h>
#include <zephyr/drivers/sensor.h>
#include <zephyr/sys/printk.h>
```

Cf https://github.com/zephyrproject-rtos/zephyr/pull/45388

erwango commented on drivers/sensor/stm32_vbat/Kconfig:
```diff
@@ -0,0 +1,10 @@
+
+config STM32_VBAT
+	bool "STM32 Vbat Sensor"
+	depends on (ADC && SOC_FAMILY_STM32)
```
```suggestion
	depends on (SOC_FAMILY_STM32 && !SOC_SERIES_STM32F1X)
	select ADC
```
Then:
- update description 
- remove `CONFIG_ADC=y` in sample

FRASTM replied:
Pb : this creates a "Dependency loop"

FRASTM commented:
Rebase on 48877f8bdcd92f7e14a72ab4c830989f15f1e884
Change path to #include <zephyr/xxx 
Change the drivers/sensor/stm32_vbat/Kconfig to `select ADC `instead of _depends on_
Consequently configuring he STM32_VBAT will also configure the ADC (no more required in the proj.conf) 
--> this has also to be changed in the **drivers/sensor/stm32_temp**

GeorgeCGV commented on drivers/sensor/stm32_vbat/stm32_vbat.c:
```diff
@@ -0,0 +1,125 @@
+
+	if (rc) {
+		LOG_DBG("Setup AIN%u got %d", data->adc_cfg.channel_id, rc);
+		return rc;
```
@FRASTM, but the mutex is not unlocked.

```suggestion
		k_mutex_unlock(&data->mutex);
		return rc;
```

FRASTM replied:
using ` goto: ` like the in the _drivers/sensor/stm32_temp/stm32_temp.c_

ycsin reviewed (dismissed):
I'd like the stm32 ADC driver to remove the voltage-divider link after the sampling of the internal channel so that it can achieve what it is intended for: prevent current drain when the vbat isn't being sampled, but that could be done in another PR.

martinjaeger commented on drivers/sensor/stm32_vbat/stm32_vbat.c:
```diff
@@ -0,0 +1,126 @@
+}
+
+static const struct stm32_vbat_config stm32_vbat_dev_config = {
+	.vbat_mv = DT_INST_PROP(0, voltage_mv), /* usually VDD_VALUE */
```
```suggestion
	.vbat_mv = DT_INST_PROP(0, voltage_mv),
```
This comment is a bit misleading. I usually try to avoid using VDD as a reference if there is a more accurate internal reference available. Also: For simple sensor nodes I would connect VDD directly to VBAT (if the voltage range allows it).

FRASTM replied:
comment removed

martinjaeger commented on dts/bindings/sensor/st,stm32-vbat.yaml:
```diff
@@ -0,0 +1,23 @@
+      default: 3
+      description: fraction of VBat to be connected to the ADC input
+
+    voltage-mv:
+      type: int
+      default: 3300
+      description: Vbat voltage reference in millivolts
```
```suggestion
    vref-mv:
      type: int
      default: 3300
      description: ADC voltage reference in millivolts
```

This voltage reference is the same for all inputs of that ADC, isn't it? Should it then not go into the ADC node?

FRASTM replied:
yes, actually it is the same as the vref_mv property  of the st,stm32-adc.yaml

martinjaeger reviewed (changes requested):
Happy to find this PR, I had already started to implement it myself.

There are a few small things that I'd like to be double-checked, though. See my comments below.

FRASTM commented on drivers/sensor/stm32_vbat/stm32_vbat.c:
```diff
@@ -0,0 +1,126 @@
+
+	voltage = data->raw * cfg->vbat_mv / 0x0FFF; /* Sensor value in millivolts */
+	/* considering the vbat input through a resistor bridge */
+	voltage = voltage * cfg->ratio;
```
right

FRASTM replied:
change to unit = Volt

FRASTM commented on drivers/sensor/stm32_vbat/stm32_vbat.c:
```diff
@@ -0,0 +1,126 @@
+};
+
+struct stm32_vbat_config {
+	int vbat_mv;
```
ok

FRASTM commented:
rebase on 801fe1fa06bd6f8cc2a55079411fd4feac55eff6
Set the Vref as a property of the ADC and get to the vbat sensor through the ADC (in the io channel)
Output the result in Volts in the samples/sensor/stm32_vbat_sensor/
Run the sample on the nucleo_wb55rg and nucleo_g071rb

martinjaeger commented on samples/sensor/stm32_vbat_sensor/sample.yaml:
```diff
@@ -0,0 +1,16 @@
+      ordered: true
+      regex:
+        - "STM32 Vbat sensor test"
+        - "Current Vbat voltage: 33(.*) mV"
```
Has to be updated to Volts.

FRASTM replied:
changed to` "Current Vbat voltage: 3(.*) V"`

martinjaeger commented on samples/sensor/stm32_vbat_sensor/boards/nucleo_g071rb.overlay:
```diff
@@ -0,0 +1,21 @@
+};
+
+&adc1 {
+    vref-mv = <3000>;
```
Nits: Use tab for indentation and not spaces.

FRASTM replied:
done

martinjaeger commented on samples/sensor/stm32_vbat_sensor/README.rst:
```diff
@@ -0,0 +1,50 @@
+
+.. code-block:: console
+
+Current Vbat voltage: 3348.0 mV
```
Has to be updated to Volts.

FRASTM replied:
done

FRASTM commented on dts/bindings/sensor/st,stm32-vbat.yaml:
```diff
@@ -0,0 +1,18 @@
+
+    ratio:
+      type: int
+      default: 3
```
could better be 1 as default value or set  the property 'required'

FRASTM committed:
boards: arm: stm32g071 nucleo defines its vref for ADC

The Vref is defined by the nucleo board hardware.
This property is common to ADC channels
and required in the DTS.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
moved: the `<vref-mv>` property of the ADC is required and defined by the board hardware

FRASTM commented:
rebase on ee22ae79a4aaa2efb90580888737257724cb5e7d

aurel32 commented on drivers/sensor/stm32_vbat/stm32_vbat.c:
```diff
@@ -0,0 +1,128 @@
+		return -ENOTSUP;
+	}
+
+	voltage = data->raw * cfg->vref_mv / 0x0FFF; /* Sensor value in millivolts */
```
Instead of adding a new `vref-mv` property, you should use `adc_ref_internal(data->adc)` instead, as this value is already defined at the ADC level.

aurel32 commented on drivers/adc/adc_stm32.c:
```diff
@@ -1145,8 +1155,10 @@ static const struct adc_stm32_cfg adc_stm32_cfg_##index = {		\
 	.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(index),			\
-	.has_temp_channel = DT_INST_PROP(index, has_temp_channel),		\
-	.has_vref_channel = DT_INST_PROP(index, has_vref_channel),		\
+	.vref_mv = DT_INST_PROP(index, vref_mv),			\
```
Sorry, I have read it to fast. You are not adding a new property, but here you adding another way to get it the ADC reference value, which was already available from the standard API.

ycsin commented:
@aurel32 could you propose a pr for the change you mentioned?
